pkg/controller/upgradeset/upgradeset.go

```diff
@@ -0,0 +1,178 @@
+	factory.BatchCreateCRDs(ctx, crd.CRD{
+		GVK:        prototype.GroupVersionKind(),
+		PluralName: apiv1.UpgradeSetResourceName,
+		Status:     true,
+		Schema:     schema,
+		Categories: []string{"all", "k3os", "upgrade"},
+		ShortNames: []string{"ug", "ugs", "upset", "upsets"},
+		//Columns: []v1beta1.CustomResourceColumnDefinition{
+		//	{
+		//		Name:     "Concurrency",
+		//		Type:     "integer",
+		//		JSONPath: ".spec.concurrency",
+		//		Format:   "int64",
+		//		Priority: 10,
+		//	},
+		//	{
+		//		Name:     "Version",
+		//		Type:     "string",
+		//		JSONPath: ".spec.version",
+		//		Priority: 10,
+		//	},
+		//},
+	})
+	return nil
+}
+
+// RegisterHandlers registers UpgradeSet handlers
+func RegisterHandlers(ctx context.Context, options Options) {
+	upgradeSets := options.ControllerFactory.K3os().V1().UpgradeSet()
+	nodeUpgrades := options.ControllerFactory.K3os().V1().NodeUpgrade()
+	nodeUpgradesApply := options.ControllerApply.WithCacheTypes(nodeUpgrades)
+
+	ctlv1.RegisterUpgradeSetGeneratingHandler(ctx, upgradeSets, nodeUpgradesApply, "", options.ControllerName,
+		generatingHandler(ctx, options), &generic.GeneratingHandlerOptions{
+			AllowClusterScoped: true,
+		},
+	)
+}
+
+// generatingHandler returns an array of NodeUpgrades based on the triggering UpgradeSet and matching Nodes
+func generatingHandler(ctx context.Context, options Options) ctlv1.UpgradeSetGeneratingHandler {
+	nodeCache := options.CoreFactory.Core().V1().Node().Cache()
+	upgradeSets := options.ControllerFactory.K3os().V1().UpgradeSet()
+
+	return func(obj *apiv1.UpgradeSet, status apiv1.UpgradeSetStatus) (objects []runtime.Object, _ apiv1.UpgradeSetStatus, _ error) {
+		var (
+			nodeNames []string
+		)
+
+		nodeSelector, err := nodeSelector(obj.Spec.Version)
+		if err != nil {
+			logrus.Error(err)
+			return objects, status, nil
+		}
+
+		if len(status.Upgrades) > 0 {
+			requirementUpgrading, err := labels.NewRequirement(corev1.LabelHostname, selection.In, status.Upgrades)
+			if err != nil {
+				logrus.Error(err)
+				return objects, status, nil
+			}
+
+			if upgradingNodes, err := nodeCache.List(nodeSelector.Add(*requirementUpgrading)); err != nil {
+				logrus.Error(err)
+			} else {
+				logrus.Debugf("upgradingNodes = %#v", upgradingNodes)
+				for _, node := range upgradingNodes {
+					//nodes[node.Name] = node
+					nodeNames = append(nodeNames, node.Name)
+				}
+			}
+
+			requirementNotUpgrading, err := labels.NewRequirement(corev1.LabelHostname, selection.NotIn, status.Upgrades)
+			if err != nil {
+				logrus.Error(err)
+				return objects, status, nil
+			}
+			nodeSelector = nodeSelector.Add(*requirementNotUpgrading)
+		}
+
+		if candidateNodes, err := nodeCache.List(nodeSelector); err != nil {
+			logrus.Error(err)
+		} else {
+			logrus.Debugf("candidateNodes = %#v", candidateNodes)
+			for i := 0; i < len(candidateNodes) && uint64(len(nodeNames)) < obj.Spec.Concurrency; i++ {
+				nodeNames = append(nodeNames, candidateNodes[i].Name)
```

Comment:
I think if there are more nodes available than the configured concurrency the generated nodeName list is not idempotent.

Reply:
I loop through and record all upgrading nodes before considering candidate nodes to guard against this. A node will fall off of the upgrading list if:
a. something other than the controller removes it from the `UpgradeSetStatus.Upgrades` list, or
b. it no longer meets the base selection criteria which in almost all cases should mean an upgrade was applied and the node's version label matches the `UpgradeSetSpec.Version`